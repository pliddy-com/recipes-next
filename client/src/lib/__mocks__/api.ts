export interface ApiTestProps {
  getNavTaxonomy: () => [{ slug: string }];
  getRecipeSlugs: () => string[];
  getTagSlugs: () => string[];
  getRecipeIndex: () => [{ slug: string }];
  getRecipeList: () => [{ slug: string }];
  getRecipePage: () => { slug: string; title: string };
  getTagIndex: () => [{ slug: string; title: string }];
}

jest.mock('urql');

// import api library to mock
const api: ApiTestProps = jest.createMockFromModule('lib/api');

api.getNavTaxonomy = jest.fn().mockResolvedValue({
  categories: [
    { slug: 'slug-1', title: 'Title 1' },
    { slug: 'slug-2', title: 'Title 2' }
  ],
  cuisine: [
    { slug: 'slug-1', title: 'Title 1' },
    { slug: 'slug-2', title: 'Title 2' }
  ],
  tags: [
    { slug: 'slug-1', title: 'Title 1' },
    { slug: 'slug-2', title: 'Title 2' }
  ]
});

api.getRecipeSlugs = jest.fn().mockResolvedValue(['slug-1', 'slug-2']);

api.getTagSlugs = jest.fn().mockResolvedValue(['slug-1', 'slug-2']);

api.getRecipeIndex = jest.fn().mockResolvedValue([
  {
    slug: 'slug-1',
    title: 'Title 1'
  },
  {
    slug: 'slug-2',
    title: 'Title 2'
  }
]);

api.getRecipeList = jest.fn().mockResolvedValue([
  {
    __typename: 'Tag',
    slug: 'baking',
    title: 'Baking',
    linkedFrom: {
      recipeCollection: {
        total: 3,
        items: [
          {
            __typename: 'Recipe',
            sys: {
              firstPublishedAt: '2022-06-17T00:02:14.528Z',
              __typename: 'Sys'
            },
            title: 'Biscuits',
            slug: 'biscuits',
            abstract:
              'Traditional baking powder biscuits that use grated frozen butter and folded dough to give the biscuits a soft, flaky texture.',
            image: {
              __typename: 'Asset',
              title: 'Biscuits',
              description:
                'A batch of baking soda biscuits on a parchment-lined baking sheet.',
              contentType: 'image/jpeg',
              fileName: 'biscuits.jpg',
              size: 2583242,
              url: 'https://images.ctfassets.net/fo9qwg6zarbt/B6C4D23gXNDNfGVrxb1DK/9d8e62c1faf660f7e542eed14e044b57/biscuits.jpg',
              height: 3024,
              width: 4032
            },
            tagsCollection: {
              items: [
                {
                  __typename: 'Tag',
                  title: 'Baking',
                  slug: 'baking'
                },
                {
                  __typename: 'Tag',
                  title: 'Biscuits',
                  slug: 'biscuits'
                },
                {
                  __typename: 'Tag',
                  title: 'Breakfast',
                  slug: 'breakfast'
                },
                {
                  __typename: 'Tag',
                  title: 'Southern',
                  slug: 'southern'
                }
              ],
              __typename: 'RecipeTagsCollection'
            }
          },
          {
            __typename: 'Recipe',
            sys: {
              firstPublishedAt: '2022-11-15T21:50:41.051Z',
              __typename: 'Sys'
            },
            title: 'No-Knead Olive Bread',
            slug: 'no-knead-olive-bread',
            abstract:
              'A savory variation of no-knead bread with the addition of olives, garlic, and rosemary to give it a Mediterranean twist.',
            image: {
              __typename: 'Asset',
              title: 'Olive Bread',
              description: 'Loaf of no-knead olive bread',
              contentType: 'image/jpeg',
              fileName: 'olive-bread.JPG',
              size: 2849709,
              url: 'https://images.ctfassets.net/fo9qwg6zarbt/2rUa4IhfkLtRRBmHS6ihcD/e953337e00d1eba9551b15e7eee18229/olive-bread.JPG',
              height: 3024,
              width: 4032
            },
            tagsCollection: {
              items: [
                {
                  __typename: 'Tag',
                  title: 'Baking',
                  slug: 'baking'
                },
                {
                  __typename: 'Tag',
                  title: 'Bread',
                  slug: 'bread'
                },
                {
                  __typename: 'Tag',
                  title: 'Olives',
                  slug: 'olives'
                },
                {
                  __typename: 'Tag',
                  title: 'Mediterranean',
                  slug: 'mediterranean'
                }
              ],
              __typename: 'RecipeTagsCollection'
            }
          },
          {
            __typename: 'Recipe',
            sys: {
              firstPublishedAt: '2022-06-18T22:42:01.107Z',
              __typename: 'Sys'
            },
            title: 'No-Knead Bread',
            slug: 'no-knead-bread',
            abstract:
              'No-knead bread using instant yeast and a dutch oven or covered ceramic bread pan for a classic, crispy crust.',
            image: {
              __typename: 'Asset',
              title: 'No Knead Bread',
              description:
                'Loaf of no-knead bread cooling on a wire baking rack.',
              contentType: 'image/jpeg',
              fileName: 'no-knead-bread.JPG',
              size: 2943075,
              url: 'https://images.ctfassets.net/fo9qwg6zarbt/6HZDSm8AK4iQHr9v2UOGpq/17569d5be00441216338c7800e5ebc2c/no-knead-bread.JPG',
              height: 3024,
              width: 4032
            },
            tagsCollection: {
              items: [
                {
                  __typename: 'Tag',
                  title: 'Baking',
                  slug: 'baking'
                },
                {
                  __typename: 'Tag',
                  title: 'Bread',
                  slug: 'bread'
                }
              ],
              __typename: 'RecipeTagsCollection'
            }
          }
        ],
        __typename: 'RecipeCollection'
      },
      __typename: 'TagLinkingCollections'
    }
  }
]);

api.getRecipePage = jest.fn().mockResolvedValue({
  recipe: {
    title: 'Recipe Title',
    slug: 'slug-1',
    __typename: 'Recipe',
    sys: { firstPublishedAt: '2022-06-26T07:24:30.384Z' },
    abstract: 'abstract',
    recipeYield: 4,
    prepTime: 15,
    cookTime: 60,
    keywords: ['keyword1', 'keyword2'],
    image: {
      sys: {
        id: 'sysid-1',
        __typename: 'Sys'
      },
      __typename: 'Asset',
      title: 'Image Title',
      description: 'Image description.',
      contentType: 'image/jpeg',
      fileName: 'filename.jpg',
      size: 99999,
      url: 'https://recipes.pliddy.com/url',
      height: 300,
      width: 400
    },
    ingredientsList: [
      {
        sectionTitle: 'Section Label 1',
        sectionItems: ['ingredient 1', 'ingredient 2']
      }
    ],
    equipment: ['equipment 1', 'equipment 2'],
    instructionsList: [
      {
        sectionTitle: 'Section Label 1',
        sectionItems: ['Instructions item 1']
      },
      {
        sectionTitle: 'Section Label 2',
        sectionItems: ['Instructions item 2', 'Instructions item 3']
      }
    ],
    notes: ['note 1'],
    tagsCollection: {
      items: [
        {
          sys: {
            id: 'sysid-4',
            __typename: 'Sys'
          },
          __typename: 'Tag',
          title: 'Tag 1',
          slug: 'tag-1'
        }
      ],
      __typename: 'RecipeTagsCollection'
    }
  },
  cuisine: [
    {
      __typename: 'Tag',
      title: 'Tag 1 Title',
      slug: 'tag-1',
      linkedFrom: {
        recipeCollection: {
          total: 2,
          items: [
            {
              __typename: 'Recipe',
              slug: 'slug-1',
              title: 'Title 2'
            },
            {
              __typename: 'Recipe',
              slug: 'slug-2',
              title: 'Title 1'
            }
          ]
        }
      }
    },
    {
      __typename: 'Tag',
      title: 'Tag 2 Title',
      slug: 'tag-2',
      linkedFrom: {
        recipeCollection: {
          total: 2,
          items: [
            {
              __typename: 'Recipe',
              slug: 'slug-3',
              title: 'Title 3'
            },
            {
              __typename: 'Recipe',
              slug: 'slug-4',
              title: 'Title 5'
            }
          ]
        }
      }
    }
  ],
  categories: [
    {
      __typename: 'Tag',
      title: 'Tag 1 Title',
      slug: 'tag-1',
      linkedFrom: {
        recipeCollection: {
          total: 2,
          items: [
            {
              __typename: 'Recipe',
              slug: 'slug-1',
              title: 'Title 2'
            },
            {
              __typename: 'Recipe',
              slug: 'slug-2',
              title: 'Title 1'
            }
          ]
        }
      }
    },
    {
      __typename: 'Tag',
      title: 'Tag 2 Title',
      slug: 'tag-2',
      linkedFrom: {
        recipeCollection: {
          total: 2,
          items: [
            {
              __typename: 'Recipe',
              slug: 'slug-3',
              title: 'Title 3'
            },
            {
              __typename: 'Recipe',
              slug: 'slug-4',
              title: 'Title 5'
            }
          ]
        }
      }
    }
  ]
});

api.getTagIndex = jest.fn().mockResolvedValue([
  {
    __typename: 'Tag',
    title: 'Tag 1 Title',
    slug: 'tag-1',
    linkedFrom: {
      recipeCollection: {
        total: 2,
        items: [
          {
            __typename: 'Recipe',
            slug: 'slug-1',
            title: 'Title 2'
          },
          {
            __typename: 'Recipe',
            slug: 'slug-2',
            title: 'Title 1'
          }
        ]
      }
    }
  },
  {
    __typename: 'Tag',
    title: 'Tag 2 Title',
    slug: 'tag-2',
    linkedFrom: {
      recipeCollection: {
        total: 2,
        items: [
          {
            __typename: 'Recipe',
            slug: 'slug-3',
            title: 'Title 3'
          },
          {
            __typename: 'Recipe',
            slug: 'slug-4',
            title: 'Title 5'
          }
        ]
      }
    }
  }
]);

module.exports = api;
