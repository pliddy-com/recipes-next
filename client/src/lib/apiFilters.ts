import {
  TagDefaultFragment,
  NavTaxonomyFragment,
  TagTaxonomyFragment,
  RecipeSlugFragment
} from '../types/queries';

import { hasValue } from './utils';

export const filterSlugs = (
  items: (RecipeSlugFragment | TagDefaultFragment | null)[]
) =>
  items &&
  items
    ?.map((child) => (child && child.slug ? child.slug : null))
    .filter(hasValue);

export const filterTagsWithRecipes = ({
  tagCollection
}: {
  tagCollection: TagTaxonomyFragment;
}) =>
  tagCollection &&
  tagCollection?.items
    .filter(hasValue)
    .filter(
      (item) =>
        item?.linkedFrom?.recipeCollection?.total &&
        item?.linkedFrom?.recipeCollection?.total > 0
    );

export const filterTaxonomyItemsWithRecipes = ({
  taxonomyCollection
}: {
  taxonomyCollection: NavTaxonomyFragment;
}) => {
  return taxonomyCollection?.items?.[0]?.childrenCollection?.items?.filter(
    (item) =>
      item &&
      (('linkedFrom' in item &&
        item.linkedFrom &&
        'recipeCollection' in item.linkedFrom && // is a tag
        item?.linkedFrom?.recipeCollection?.total &&
        item?.linkedFrom?.recipeCollection?.total > 0) ||
        ('childrenCollection' in item && // is a taxonomy
          item?.childrenCollection?.total &&
          item?.childrenCollection?.total > 0))
  );
};

// TODO: update to filter out childrenCollection.items.linkedFrom.recipeCollection.total <== 0
