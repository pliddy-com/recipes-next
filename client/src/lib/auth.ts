/* istanbul ignore file */

import { AuthenticationDetails, CognitoUser } from 'amazon-cognito-identity-js';

import userPool from './userPool';

const APP_CLIENT_ID = process.env.NEXT_PUBLIC_COGNITO_APP_CLIENT_ID;
const USER_POOL_ID = process.env.NEXT_PUBLIC_COGNITO_USER_POOL_ID;
// const REGION = process.env.NEXT_PUBLIC_COGNITO_REGION;

interface signInProps {
  email: string;
  password: string;
}

export const signIn = async ({ email, password }: signInProps) => {
  if (!(email && password && APP_CLIENT_ID && USER_POOL_ID))
    throw new Error('Must provide email and password to sign in');
  try {
    // const poolData = {
    //   UserPoolId: USER_POOL_ID,
    //   ClientId: APP_CLIENT_ID
    // };

    // const userPool = new CognitoUserPool(poolData);

    const user = new CognitoUser({
      Username: email,
      Pool: userPool
    });

    const authDetails = new AuthenticationDetails({
      Username: email,
      Password: password
    });

    const res = user.authenticateUser(authDetails, {
      onSuccess: (result) => {
        console.log('login success', result);
      },
      onFailure: (err) => {
        console.log('login failure', err);
      },
      newPasswordRequired: (data) => {
        console.log('new password required', data);
      }
    });

    console.log({ res });
  } catch (e) {
    console.error(e);
    throw e;
  }
};
