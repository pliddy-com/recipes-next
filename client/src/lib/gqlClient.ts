/* istanbul ignore file */
import { cacheExchange, Client, fetchExchange, TypedDocumentNode } from 'urql';
import { ResultOf, VariablesOf } from '@graphql-typed-document-node/core';

const ACCESS_TOKEN = process.env.NEXT_PUBLIC_CONTENTFUL_ACCESS_TOKEN;
const API_ENDPOINT = process.env.NEXT_PUBLIC_CONTENTFUL_CONTENT_API;

export const gqlClient = new Client({
  url: API_ENDPOINT || '',
  exchanges: [cacheExchange, fetchExchange],
  fetchOptions: () => ({
    headers: { authorization: ACCESS_TOKEN ? `Bearer ${ACCESS_TOKEN}` : '' }
  })
});

export async function queryGraphQLContent<
  TDoc extends TypedDocumentNode<ResultOf<TDoc>, VariablesOf<TDoc>>,
  TVars = TDoc extends unknown ? never : VariablesOf<TDoc>
>(document: TDoc, variables?: TVars) {
  const { data, error } = await gqlClient
    // @ts-expect-error query variables are typed by graphQLRequest
    .query(document, variables)
    .toPromise();

  if (!data) {
    if (error) {
      throw error;
    }
    throw new Error('No data returned');
  }

  return data;
}
