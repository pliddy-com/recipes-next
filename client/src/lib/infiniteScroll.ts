/**
 *  getPagedRecipes parsing function (parse in realtime instead of pre-rendering array of arrays?)
 */

export interface PaginateResultsProps {
  data: (object | null)[];
  pageSize: number;
}

export const paginateResults = ({ data, pageSize }: PaginateResultsProps) => {
  const numPages = Math.ceil(data.length / pageSize);
  const results = [];

  for (let pageNum = 0; pageNum < numPages; pageNum++) {
    const startItem = pageNum * pageSize;
    const pageData = data.slice(startItem, startItem + pageSize);
    results.push(pageData);
  }

  return results;
};

/**
 *  LoadNext function for infinite scrolling recipes
 */

export interface LoadNextProps {
  pagedResults: (object | null)[][];
  pageNum: number;
  pageSize: number;
  setDataCallback: (pagedResults: (object | null)[][]) => void;
  setPageNumCallback: (num: number) => void;
}

export const loadNext = ({
  pagedResults,
  pageNum,
  setDataCallback,
  setPageNumCallback
}: LoadNextProps): void => {
  if (pageNum < pagedResults.length) setPageNumCallback(pageNum + 1);
  setDataCallback(pagedResults.slice(0, pageNum + 1));
};
