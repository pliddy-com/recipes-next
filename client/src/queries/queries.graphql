# query for Taxonomy with Category & Subcategory levels
# getNavTaxonomy

query navMenuData {
  categories: taxonomyCollection(
    where: { slug: "categories" }
    order: title_ASC
    limit: 1
  ) {
    ...NavTaxonomy
  }

  cuisine: taxonomyCollection(
    where: { slug: "cuisine" }
    order: title_ASC
    limit: 1
  ) {
    ...NavTaxonomy
  }

  tags: tagCollection(order: title_ASC) {
    ...TagTaxonomy
  }
}

# Query for recipe slugs
# getRecipeSlugs

query recipeSlugs($where: RecipeFilter) {
  recipeCollection(where: $where, order: title_ASC) {
    __typename
    items {
      ...RecipeSlug
    }
  }
}

# query for category slugs
# getTagSlugs

query tagSlugs($where: TagFilter) {
  tagCollection(where: $where, order: title_ASC) {
    ...TagTaxonomy
  }
}

# query for lists of linked entries for tags and images
# getLinkedEntries

query linkedEntries {
  tags: tagCollection(order: title_ASC) {
    items {
      ...TagDefault
    }
  }
  images: assetCollection(order: fileName_ASC) {
    items {
      ...ImageDefault
    }
  }
}

# Query for List Page RecipeCollection (all recipes matching tag)
# getRecipeList

query recipeList($where: TagFilter) {
  __typename
  tagCollection(where: $where, order: title_ASC, limit: 1) {
    __typename
    total
    items {
      ...ListPageItem
    }
  }
}

# Query for Recipe Collection for home (index) page
# getRecipeIndex

query recipeIndex($where: RecipeFilter) {
  __typename
  recipeCollection(where: $where, order: title_ASC) {
    total
    items {
      ...RecipeSummary
    }
  }
}

# Query for Recipe page
# getRecipePage

query recipePage($where: RecipeFilter) {
  recipeCollection(where: $where, order: title_ASC, limit: 1) {
    __typename
    items {
      ...RecipeDefault
    }
  }

  categoriesTaxonomy: taxonomyCollection(
    where: { slug: "categories" }
    order: title_ASC
    limit: 1
  ) {
    ...SchemaTaxonomy
  }

  cuisineTaxonomy: taxonomyCollection(
    where: { slug: "cuisine" }
    order: title_ASC
    limit: 1
  ) {
    ...SchemaTaxonomy
  }
}

query recipeEntryQuery($id: String!) {
  recipe(id: $id) {
    sys {
      id
      publishedVersion
    }

    ingredientsList
    instructionsList
  }
}

# used for recipe content structure migration
query recipeCollectionQuery {
  recipeCollection {
    total
    items {
      sys {
        id
        publishedVersion
      }

      ingredientsList
      instructionsList
    }
  }
}
