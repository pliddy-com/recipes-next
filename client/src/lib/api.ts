/* istanbul ignore file */

import { queryGraphQLContent } from './gqlClient';
// import * as fs from 'fs';

import {
  filterSlugs,
  filterTagsWithRecipes,
  filterTaxonomyItemsWithRecipes
} from './apiFilters';

import {
  NavMenuDataDocument,
  RecipeIndexDocument,
  RecipeIndexQueryVariables,
  RecipeListDocument,
  RecipeListQueryVariables,
  RecipePageDocument,
  RecipePageQueryVariables,
  RecipeSlugsDocument,
  RecipeSlugsQueryVariables,
  TagSlugsDocument,
  TagSlugsQueryVariables,
  NavMenuDataQuery,
  RecipeSlugsQuery,
  TagSlugsQuery,
  RecipeListQuery,
  LinkedEntriesQuery,
  LinkedEntriesDocument
} from '../types/queries';

import { hasValue } from './utils';

// const ACCESS_TOKEN = process.env.NEXT_PUBLIC_CONTENTFUL_MANAGEMENT_TOKEN ?? '';
// const SPACE_ID = process.env.NEXT_PUBLIC_CONTENTFUL_SPACE_ID ?? '';
// const API_ENDPOINT = process.env.NEXT_PUBLIC_CONTENTFUL_CONTENT_API;

// const client = contentful.createClient({
//   accessToken: ACCESS_TOKEN,
//   space: SPACE_ID
// });

// used to query content for nav menu taxonomy on client layout
export const getNavTaxonomy = async () => {
  const { queryGraphQLContent } = await import('lib/gqlClient');

  const results: NavMenuDataQuery = await queryGraphQLContent(
    NavMenuDataDocument
  );

  const { categories, cuisine, tags } = results;

  return {
    ...(categories && {
      categories: filterTaxonomyItemsWithRecipes({
        taxonomyCollection: categories
      })
    }),
    ...(cuisine && {
      cuisine: filterTaxonomyItemsWithRecipes({ taxonomyCollection: cuisine })
    }),
    ...(tags && {
      tags: filterTagsWithRecipes({ tagCollection: tags })
    })
  };
};

// used to get linked entries for recipe edit

export const getLinkedEntries = async () => {
  const { queryGraphQLContent } = await import('lib/gqlClient');

  const results: LinkedEntriesQuery = await queryGraphQLContent(
    LinkedEntriesDocument
  );

  const { images, tags } = results;

  return {
    ...(images && {
      images: images.items
    }),
    ...(tags && {
      tags: tags.items
    })
  };
};

// used by getStaticProps for recipe page
export const getRecipeSlugs = async (variables: RecipeSlugsQueryVariables) => {
  const results: RecipeSlugsQuery = await queryGraphQLContent(
    RecipeSlugsDocument,
    variables
  );

  const { recipeCollection } = results;

  return recipeCollection && recipeCollection.items
    ? filterSlugs(recipeCollection.items)
    : [];
};

// used by getStaticProps for tag page
export const getTagSlugs = async (variables?: TagSlugsQueryVariables) => {
  const results: TagSlugsQuery = await queryGraphQLContent(
    TagSlugsDocument,
    variables
  );

  const { tagCollection } = results;

  const tags =
    tagCollection &&
    filterTagsWithRecipes({
      tagCollection
    });

  return tags ? filterSlugs(tags) : [];
};

// used to query content for home index page
export const getRecipeIndex = async (variables?: RecipeIndexQueryVariables) => {
  const { recipeCollection } = await queryGraphQLContent(
    RecipeIndexDocument,
    variables
  );

  return recipeCollection ? recipeCollection.items.filter(hasValue) : [];
};

// used to query content for tag & category page
export const getRecipeList = async (variables?: RecipeListQueryVariables) => {
  const { tagCollection } = await queryGraphQLContent(
    RecipeListDocument,
    variables
  );

  return tagCollection ? tagCollection.items.filter(hasValue) : [];
};

// used to query content for standalone recipe page
export const getRecipePage = async (variables?: RecipePageQueryVariables) => {
  const { recipeCollection, categoriesTaxonomy, cuisineTaxonomy } =
    await queryGraphQLContent(RecipePageDocument, variables);

  const categories =
    categoriesTaxonomy?.items.filter(hasValue)[0].childrenCollection?.items;

  const cuisine =
    cuisineTaxonomy?.items.filter(hasValue)[0].childrenCollection?.items;

  const recipe = recipeCollection?.items.filter(hasValue)[0];

  return {
    ...(recipe && { recipe }),
    ...(categories && { categories }),
    ...(cuisine && { cuisine })
  };
};

// used to query content for tag index page
export const getTagIndex = async (variables?: RecipeListQueryVariables) => {
  const results: RecipeListQuery = await queryGraphQLContent(
    RecipeListDocument,
    variables
  );

  const { tagCollection } = results;

  return tagCollection
    ? filterTagsWithRecipes({
        tagCollection
      })
    : [];
};
