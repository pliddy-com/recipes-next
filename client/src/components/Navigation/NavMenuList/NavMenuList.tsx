import { MouseEventHandler } from 'react';

import Link from 'next/link';

import Divider from '@mui/material/Divider';
import ListItemButton from '@mui/material/ListItemButton';
import ListItemText from '@mui/material/ListItemText';
import MenuItem from '@mui/material/MenuItem';
import MenuList from '@mui/material/MenuList';

import CategoryMenu from 'components/Navigation/NavMenu/CategoryMenu/CategoryMenu';

import { TaxonomyChildrenItem } from 'types/queries';
import ListSubheader from '@mui/material/ListSubheader';

export interface NavMenuListProps {
  className: string;
  featuredLabel?: string;
  featuredUrl?: string;
  id: string;
  label: string;
  onClick: MouseEventHandler;
  nav: TaxonomyChildrenItem[];
  root: string;
  showLabel?: boolean;
}

const NavMenuList = ({
  className,
  featuredLabel,
  featuredUrl,
  id,
  label,
  nav,
  onClick,
  root,
  showLabel
}: NavMenuListProps) => {
  return (
    <MenuList
      aria-label={`${id} menu`}
      className={className}
      role="menu"
      component="nav"
      {...(showLabel && { subheader: <ListSubheader>{label}</ListSubheader> })}
    >
      {featuredLabel && featuredUrl && (
        <MenuItem className="menuItem featured">
          <ListItemButton component={Link} href={featuredUrl} onClick={onClick}>
            <ListItemText primary={featuredLabel} />
          </ListItemButton>
          <Divider />
        </MenuItem>
      )}
      {nav &&
        nav.map((item) => {
          // if includes 'tag', it's a Taxonomy, else it's a Tag
          const categoryTag =
            item && item.__typename === 'Taxonomy' ? item.tag : item;

          const { linkedFrom } = categoryTag ?? {};

          return linkedFrom && 'recipeCollection' in linkedFrom ? (
            <CategoryMenu
              key={item?.slug}
              category={item}
              onClick={onClick}
              root={root}
            />
          ) : null;
        })}
    </MenuList>
  );
};

export default NavMenuList;
