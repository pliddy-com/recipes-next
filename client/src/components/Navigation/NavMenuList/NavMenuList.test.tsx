// import testing-library methods
import { render } from '@testing-library/react';

// add custom jest matchers from jest-dom
import '@testing-library/jest-dom';

import NavMenuList, { NavMenuListProps } from './NavMenuList';

import { TaxonomyChildrenItem } from 'types/queries';

import * as api from 'lib/api';

jest.mock('lib/api');

const callback = jest.fn();

describe('NavMenuList', () => {
  describe('when there is a properly structured nav property', () => {
    it('it renders a nav menu list', async () => {
      const { categories } = await api.getNavTaxonomy();

      const props: NavMenuListProps = {
        className: 'menuList',
        featuredLabel: 'Categories',
        featuredUrl: '/tags',
        id: 'categories',
        label: 'Categories',
        onClick: callback,
        nav: categories as TaxonomyChildrenItem[],
        root: 'tag',
        showLabel: true
      };

      const { asFragment } = render(<NavMenuList {...props} />);

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when there are not properly structured category tags', () => {
    const nav = [
      {
        __typename: 'Tag',
        title: 'Tag A',
        slug: 'tag-a',
        linkedFrom: null
      },
      {
        __typename: 'Tag',
        title: 'Tag B',
        slug: 'tag-b',
        linkedFrom: {
          recipeCollection: null
        }
      },
      {
        __typename: 'Tag',
        title: 'Tag C',
        slug: 'tag-c'
      },
      {
        __typename: 'Taxonomy',
        title: 'Tag D',
        slug: 'tag-d',
        linkedFrom: {
          recipeCollection: null
        }
      }
    ];

    it('it does not render', () => {
      const props: NavMenuListProps = {
        className: 'menuList',
        featuredLabel: 'Categories',
        featuredUrl: '/tags',
        id: 'categories',
        label: 'Categores',
        onClick: callback,
        nav: nav as TaxonomyChildrenItem[],
        root: 'tag',
        showLabel: true
      };

      const { container } = render(<NavMenuList {...props} />);

      expect(container.getElementsByClassName('menu').length).toEqual(0);
    });
  });
});
