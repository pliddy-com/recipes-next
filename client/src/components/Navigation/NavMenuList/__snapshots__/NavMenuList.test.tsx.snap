// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavMenuList when there is a properly structured nav property it renders a nav menu list 1`] = `
<DocumentFragment>
  <nav
    aria-label="categories menu"
    class="MuiList-root MuiList-padding MuiList-subheader menuList css-p64g7a-MuiList-root"
    role="menu"
    tabindex="-1"
  >
    <li
      class="MuiListSubheader-root MuiListSubheader-gutters MuiListSubheader-sticky css-uob957-MuiListSubheader-root"
    >
      Categories
    </li>
    <div
      class="MuiBox-root css-0"
      tabindex="0"
    >
      <li
        class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters menuItem featured css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
        role="menuitem"
        tabindex="-1"
      >
        <a
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
          href="/tags"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root css-tlelie-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              Categories
            </span>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </a>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </li>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
        style="margin-top: 0px; margin-bottom: 0px;"
      />
    </div>
  </nav>
</DocumentFragment>
`;
